### PR TITLE
administration/upgrading: upgrade guide

### PR DESCRIPTION
**Pull request by bboozzoo**
@mendersoftware/rndity @kacf @maciejmrowiec @pasinskim @GregorioDiStefano 

BTW. no clue if there's anything needed from client side. Feel free to open a separate PR with client related bits.

### CODE REVIEW

File: 06.Administration/06.Upgrading/docs.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+
+As a good engineering practice, it is advisable to perform the upgrade on a
+staging environment first. This will allow you to discover potential problems
+and allow to exercise the procedure in a safe manner. 
+
+! Upgrade procedure involves some downtime.
+
+Production installation is largely based on using git and Mender integration
+repository. This is the reason that the upgrade procedure follows a regular git
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: This is the reason s/that/why/

File: 06.Administration/06.Upgrading/docs.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+workflow with branching, pulling remote changes and merging locally.
+
+## Backing up existing data
+
+Before an upgrade will be applied it is advisable to backup existing data and
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: Maybe: `Before an upgrade will be performed`?

File: 06.Administration/06.Upgrading/docs.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+volumes.
+
+Consult Mongo DB and docker manuals for the necessary steps.
+
+## Updating your local repository
+
+The first step is to upgrade you local repository by pulling changes from Mender
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: ...is to upgrade s/you/your/

File: 06.Administration/06.Upgrading/docs.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+```
+> Fetching origin
+> remote: Counting objects: 367, done.
+> remote: Compressing objects: 100% (31/31), done.
+> remote: Total 367 (delta 134), reused 122 (delta 122), pack-reused 214
+> Receiving objects: 100% (367/367), 83.55 KiB | 0 bytes/s, done.
+> Resolving deltas: 100% (214/214), completed with 42 local objects.
+> From https://github.com/mendersoftware/integration
+>  * 02cd118..75b7831      1.0.x      -> origin/1.0.x
+>  * [new branch]      GregorioDiStefano-patch-1 -> origin/GregorioDiStefano-patch-1
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: Do we need this?

File: 06.Administration/06.Upgrading/docs.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+>    06f3212..e9e5df4  master     -> origin/master
+>  * [new branch]      revert-152-cert-default-keygen -> origin/revert-152-cert-default-keygen
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: And this?

File: 06.Administration/06.Upgrading/docs.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+For each release, there will be a corresponding release branch. For example, a
+branch named `1.0.x`, provides a 1.0 release setup. Recall
+from [Production installation](../production-installation) guide that our local
+setup was introduced in a branch that was created from some release version.
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: Maybe `...that was created from s/some/given/ release version`?

File: 06.Administration/06.Upgrading/docs.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+You can use git commands such as `git log`, `git diff` to review the changes
+introduced in upstream branch. For example:
+
+```bash
+git log HEAD..origin/1.0.x
+```
+
+The most important thing to review is a diff in production template between our
+version and the version present in repository. For patch release version there
```
Comment (pasinskim) [inline 06.Administration/06.Upgrading/docs.md]: For the patch release version...

Comment (bboozzoo) [inline 06.Administration/06.Upgrading/docs.md]: fixed

File: 06.Administration/06.Upgrading/docs.md (hunk 6 above)
Comment (bboozzoo) [inline 06.Administration/06.Upgrading/docs.md]: fixed

File: 06.Administration/06.Upgrading/docs.md (hunk 5 above)
Comment (bboozzoo) [inline 06.Administration/06.Upgrading/docs.md]: removed both

File: 06.Administration/06.Upgrading/docs.md (hunk 3 above)
Comment (bboozzoo) [inline 06.Administration/06.Upgrading/docs.md]: fixed

File: 06.Administration/06.Upgrading/docs.md (hunk 2 above)
Comment (bboozzoo) [inline 06.Administration/06.Upgrading/docs.md]: fixed

File: 06.Administration/06.Upgrading/docs.md (hunk 1 above)
Comment (bboozzoo) [inline 06.Administration/06.Upgrading/docs.md]: fixed

**Comment by maciejmrowiec** [comment]
Nice guide, one catch is that we should instruct people to upgrade to specific or latest release in contrary to asking them to install unreleased latest commits (1.0.x).

**Comment by bboozzoo** [comment]
One thing to keep in mind is that when 1.0.1 is released, you'll just `git merge 1.0.1`. Another thing is that 1.0.1 tag is not present there yet. 

Anyways, I think the upgrade guide will need an update just before we release. I suppose this, along with upgrade testing, should be accounted for in relevant production upgrade tasks (no clue if we have these already or not).

**Comment by maciejmrowiec** [comment]
as long of is part of https://tracker.mender.io/browse/MEN-1112

**Comment by maciejmrowiec** [comment]
No need for a new task, https://tracker.mender.io/browse/MEN-1112 specifies exactly that. Tomorrow @michaelatmender should be around to help with formatting issue.

**Comment by bboozzoo** [comment]
Pushed another commit that pretends that there's 1.0.1 tag in remote repository. Since there's obviously no such tag in the repo, command output may not look exactly like it will when the tag appears.

@maciejmrowiec MEN-1112 also mentions client stuff. Can't tell if there's anything changed there, perhaps @kacf or @pasinskim Feel free to reassign the task to yourself

**Comment by kacf** [comment]
I don't think anything is required in the client. No storage of data has changed.

**Comment by bboozzoo** [comment]
I've pushed some additional patches that address formatting. I've finally figured out the formatting and why it was broken initially. Turns out we are using quotation and line breaks (the non obvious two-or-more whitespace at the end of the line) to display command output. This is broken for two reasons:
- the output is no longer preformatted (any leading whitespace in command output will be lost, spaces between will be lost etc)
- quoted text is still interpreted as markdown, so any valid markdown tags will trigger corresponding formatting in the output (ex. `#!/bin/bash` yields a top level header `!/bin/bash`, or `* [new tag] ...` in git fetch output will produce list elements)

**Comment by michaelclelland** [comment]
In markdown you can escape characters like * and # by using a backslash: https://daringfireball.net/projects/markdown/syntax#backslash

If you want to preserve leading white space for indenting lines, you can use `&nbsp;`

**Comment by michaelatmender** [comment]
In markdown you can escape characters like * and # by using a backslash: https://daringfireball.net/projects/markdown/syntax#backslash

If you want to preserve leading white space for indenting lines, you can use &nbsp;

**Comment by bboozzoo** [comment]
Merge, no merge?

**Comment by maciejmrowiec** [comment]
if formatting issues are gone, lets merge

**Pull request merged**